### ISSUE TITLE
Standardize alternate spelling of "class"

### ISSUE THREAD

**Issue by selik** [pandas-dev/pandas#21017]
I'd like to write `cls` instead of `klass` when creating a generic variable referring to a class object. Both are somewhat standard (mis)spellings across the Python community, but `cls` is a bit more popular, judging from my non-systematic survey.

Sadly, I think `klass` has some awkward cultural undertones, due to its similarity to the word "klan". I understand that's not a global truth, but it leaves a bad taste in my mouth. Making that connection probably wouldn't occur to most folks, but I can't escape it. Happily, we don't have to argue about it as the Pandas project uses both terms and it'd be just as easy to standardize on `cls` instead.

Further, SublimeText's default Python syntax highlighter recognizes `cls` but not `klass`. I'm betting some other editors do the same, changing `cls`'s styling to match `self`.

**Comment by WillAyd** [comment]
I don't have a strong preference on convention here but do you have other projects you can point to that use this? `cls` stylistically from PEP8 is preferred as the first argument to class methods, which is conceptually different from using as a parameter in pytest.

Not familiar with SublimeText but I'd imagine it's syntax highlighting is intended for those class methods and not related to parametrization. In a similar art we wouldn't (I don't think) ever use `self` as a parameter name.

Also wondering if we'd ever have a parametrized class setup method where the usage of `cls` would conflict with the intended usage of it as the first argument. Admittedly a very obscure use case but ideally want a standard that would work in all instances.

**Comment by selik** [comment]
Interesting point. Searching the CPython repository shows at least two dozen occurrences of "cls" used as a parameter outside of classmethods. There are about half as many occurrences of "klass" in the same manner, none of which are motivated by the use of a classmethod.

The variable name "self" has a more special-purpose meaning. We're lucky that there are other terms like "obj" or "instance" that could mean the same thing. For a class, we don't have the variety of synonyms. There's "type", but that's also burdened by being a builtin.

The special case you described sounds plausible, but I can't remember any time I've used that pattern in the last 10 years.

**Comment by shoyer** [comment]
I also like `cls` more than `klass`. `class_` is also a reasonable alternative.

**Comment by toobaz** [comment]
> Anyway, here's my evidence that "cls" is the right choice:

Doesn't this disregard the possible ambiguity of ``cls``?

I find ``cls`` OK aestethically, but I think distinguishing it from the parameter in classmethods is more important.

And while I'm no particular fan of ``klass``, I definitely like it more than ``class_`` (I think in Python we tend to attach meaning to underscores and not to "k"s). I would immediately change preferences if I was provided clues that the widespread use of ``klass`` is an intended reference to white supremacist groups, but I'm aware of none. On the other hand, there are other misspellings availables: for instance I don't see downsides (if we omit the fact that ``klass`` is somewhat standard outside pandas) in ``clas``.

**Comment by TomAugspurger** [comment]
> I think in Python we tend to attach meaning to underscores and not to "k"s

FWIW, PEP8 recommends a trailing underscore specifically for this case: https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles (avoiding a conflict with a keyword / builtin)

**Comment by selik** [comment]
@toobaz As far as I know, there's no malicious intent in `klass`. I happen to have grown up near a Confederate memorial (https://en.wikipedia.org/wiki/Stone_Mountain#Confederate_Memorial), so these accidental similarities probably bother me more than others.

Maybe I should just get over it, but since this project uses both, I thought I might be able to flip it all to just one standard.

**Comment by selik** [comment]
> Doesn't this disregard the possible ambiguity of cls?

@toobaz That was in response to the question of whether that ambiguity matters, demonstrating that `cls` is often used outside of classmethods.

**Comment by toobaz** [comment]
> @toobaz That was in response to the question of whether that ambiguity matters, demonstrating that cls is often used outside of classmethods in CPython and the standard library.

Sorry, I must have missed this... how do you show that your count refers to the use "_outside of classmethods_"?

**Comment by selik** [comment]
The grep was for `", cls"` not `"(cls"`. A classmethod would have been the latter. Though I suppose I didn't check for if `klass` was the first parameter as the typical style is `cls` as the first parameter for classmethods.

**Comment by toobaz** [comment]
> The grep was for ", cls" not "(cls". A classmethod would have been the latter

would have been _also_ the latter... but you typically do something with the ``cls`` parameter of a classmethod... and
 
```
$ grep -r "(cls, " cpython-master | wc -l
685
```

(not that this proves that ``cls`` is necessarily a bad choice, of course)

**Comment by selik** [comment]
I think I've discovered the *de facto* standard from CPython is to use "type" as the 2nd class-object parameter name when "cls" is the first parameter and a second class-object parameter is desired. For example, EnumMeta has `def _create_(cls, class_name, names, *, module=None, qualname=None, type=None, start=1)`.

Interestingly, it seems slightly more common to receive a tuple of classes than a single class as a 2nd parameter, so name collisions is avoided by saying "types" or "bases".

All this discussion and I think I've talked myself down from the argument. Sorry for the bikeshedding. I still prefer "cls" though :-)
